backend/routes/forms/submit.py

```diff
@@ -168,16 +168,45 @@ async def submit(self, request: Request) -> JSONResponse:
             if any("unittests" in question.data for question in form.questions):
                 unittest_results = await execute_unittest(response_obj, form)
 
-                if not all(test.passed for test in unittest_results):
-                    # Return 500 if we encountered an internal error (code 99).
-                    status_code = 500 if any(
-                        test.return_code == 99 for test in unittest_results
-                    ) else 403
+                failures = []
+                status_code = 403
 
+                for test in unittest_results:
+                    response_obj.response[test.question_id] = {
+                        "value": response_obj.response[test.question_id],
+                        "passed": test.passed
+                    }
+
+                    if test.return_code == 0:
+                        failure_names = [] if test.passed else test.result.split(";")
+                    elif test.return_code == 5:
+                        failure_names = ["Could not parse user code."]
+                    elif test.return_code == 6:
+                        failure_names = ["Could not load user code."]
+                    else:
+                        failure_names = ["Internal error."]
+
+                    response_obj.response[test.question_id]["failures"] = failure_names
+
+                    # Report a failure on internal errors,
+                    # or if the test suite doesn't allow failures
+                    if not test.passed:
+                        allow_failure = (
+                            form.questions[test.question_index].data["unittests"]["allow_failure"]
+                        )
+
+                        if test.return_code == 99:
```

Comment:
Might be worth documenting when return code 99 happens.
```suggestion
                        # An error while communicating with the test runner
                        if test.return_code == 99:
```